### PR TITLE
Port two Endomorphism submodules over to new Function hierarchy

### LINKED ISSUE #2299: Update `Function.Endomorphism.*` to avoid use of deprecated `Function.Equality`
The #759 deprecation effort left some residues which still import (at least) `Function.Equality` (with a pragma to suppress the warning), notably:
- [x] `Function.Endomorphism.Setoid`
- [x] `Function.Endomorphism.Propositional`

which define the monoid etc. of endomorphisms (needed for [a development of `Foldable`](https://github.com/agda/agda-stdlib/pull/2300) cf. generalising #1281 , and also now #2350 ). 

UPDATED: ~~blocked on not-yet-merged #2240~~ each of the above also uses the long-since-deprecated (v1.5!) `Algebra.Morphism.Is*Morphism` definitions instead of updating to the ones in `Algebra.Morphism.Structures` ...

Should update to remove dependency on the deprecated module(s).

### PR DESCRIPTION
**Pull request by JacquesCarette**
Closes #2299 

Needed to add some missing declarations in `Algebra.Morphism.Structures` to make it all go through.

### CODE REVIEW

File: src/Algebra/Morphism/Structures.agda
Comment (MatthewDaggitt) [inline src/Algebra/Morphism/Structures.agda:72]: I'm slightly reluctant to go down this path, as this opens up to us adding a crazy number of synonyms for *all* the structures/bundles in `Algebra.Bundles`. This is the same reason why we don't have `RawSemigroup`...

**Review by MatthewDaggitt — CHANGES_REQUESTED** [review]
Gah, I can't believe I missed these.

Unfortunately, I hate to sound like a broken record but these are breaking changes. I know that *I* messed up in v2.0 by not migrating these across, but that doesn't necessarily mean that we can do it v2.1.

As a fair few names are changing, It looks like some of this can be done in a backwards compatible manner by adding deprecation warnings to the old names (e.g. `^-isSemigroupMorphism`).

Can we separate this PR out into a two parts, a set of backwards compatible changes, and a set of non-backwards compatible changes and then we can evaluate from there?

File: src/Algebra/Morphism/Structures.agda
Comment (JacquesCarette) [reply src/Algebra/Morphism/Structures.agda:72]: So the funny reason for adding this one is that the *deprecation* warning for `IsSemingroupMorphism` mentioned `IsSemigroupHomomorphism`... even though it didn't in fact exist!

Thus I decided to make the deprecation warning less of a lie.

I'm fine with not adding a crazy number of synonyms.

Comment (MatthewDaggitt) [reply src/Algebra/Morphism/Structures.agda:72]: Ah my fault again. I guess we need to update the deprecation warning to `IsMagmaHomomorphism`...

**Comment by JacquesCarette** [comment]
> Can we separate this PR out into a two parts, a set of backwards compatible changes, and a set of non-backwards compatible changes and then we can evaluate from there?

Sure. But I've never worried much about backwards compatibility, so I'm entirely sure which ones those are! I'd appreciate a little bit of guidance, and then I will do that.

File: src/Algebra/Morphism/Structures.agda
Comment (JacquesCarette) [reply src/Algebra/Morphism/Structures.agda:72]: Done in PR #2344 . I'll also 'fix' this PR in that way, though I know I'll still need to split it.

**Comment by jamesmckinna** [comment]
> Gah, I can't believe I missed these.

I think it's not so surprising, given that explicit pragmas had been inserted precisely to suppress the deprecation warnings... (hint: sometimes necessary, but probably harmful overall...)

In any case, the push to close #759 before v2.0 had been sufficiently effortful that there were bound to be 'loose ends'. 

This one, in any case, only emerged when I started wanting to thinking about `Endo` at the `Setoid` level... I think it often requires change in one place to 'provoke' reconsideration of older decisions, so we shouldn't feel too bad about that?

**Comment by JacquesCarette** [comment]
The question remains: what should this PR actually look like? I'm having a hard time seeing the 'two parts' clearly, for some reason.

And yeah, there's a very high cost to having pragmas to suppress deprecation warnings. I would be more tempted to filter them out post-facto rather than with 'global' pragmas. [If we had more precise pragmas, that would work too.]

File: CHANGELOG.md
Comment (jamesmckinna) [inline CHANGELOG.md:183]: This looks like spurious whitespace?

File: src/Function/Endomorphism/Propositional.agda
Comment (jamesmckinna) [inline src/Function/Endomorphism/Propositional.agda:101]: It's not clear to me why this name is being retained, rather than being deprecated in favour of:
```agda
^-isMagmaHomomorphism : ∀ f → IsMagmaHomomorphism +-rawMagma ∘-rawMagma (f ^_)
```

File: src/Function/Endomorphism/Setoid.agda
Comment (jamesmckinna) [inline src/Function/Endomorphism/Setoid.agda:33]: Given the semantics are the same, IIUC, why not telescope this into:
```agda
open module E = Setoid (S ⇨ S) hiding (refl)
```

File: src/Function/Endomorphism/Setoid.agda
Comment (jamesmckinna) [inline src/Function/Endomorphism/Setoid.agda:68]: Suggest adding this to the `private` block L43-45 and then using `S.refl` etc. subsequently throughout?

File: src/Function/Endomorphism/Propositional.agda
Comment (jamesmckinna) [reply src/Function/Endomorphism/Propositional.agda:101]: Also: can this be made an instance of the corresponding construction in `Setoid`?

File: CHANGELOG.md
Comment (jamesmckinna) [inline CHANGELOG.md:44]: Nice addition to the `CHANGELOG` rubric... and definitely preferable to inventing bogus 'alternative' names...?

**Comment by jamesmckinna** [comment]
Re: @MatthewDaggitt 's "non-backwards compatible", I'm having trouble seeing what is such about @JacquesCarette 's "undeprecation" strategy. Or is something else going on that I have missed?

File: src/Function/Endomorphism/Propositional.agda
Comment (jamesmckinna) [inline src/Function/Endomorphism/Propositional.agda:17]: Why promote `_+_` in the `using` list? Surely it 'follows' `zero` and `suc` in conceptual priority, as originally?

File: src/Function/Endomorphism/Setoid.agda
Comment (jamesmckinna) [inline src/Function/Endomorphism/Setoid.agda:21]: This feels a bit 'occult' to me, given that you do *not* explicitly write `using (... module ℕ ...)` on the line above? And is this solely for the sake of the constructors?

Comment (JacquesCarette) [reply src/Function/Endomorphism/Setoid.agda:21]: Note how this occult practice was already in the original, I just made it clearer by inserting a line break!

However, I should indeed edit this to remove such shenanigans, while I'm "in" the file anyways.

File: src/Function/Endomorphism/Propositional.agda
Comment (JacquesCarette) [reply src/Function/Endomorphism/Propositional.agda:17]: I have no idea why I did that. Will revert.

File: CHANGELOG.md
Comment (JacquesCarette) [reply CHANGELOG.md:183]: Yup - fixed.

File: src/Function/Endomorphism/Setoid.agda
Comment (JacquesCarette) [reply src/Function/Endomorphism/Setoid.agda:33]: Good catch - done.

File: src/Function/Endomorphism/Setoid.agda
Comment (JacquesCarette) [reply src/Function/Endomorphism/Setoid.agda:68]: Yes, I should have spotted that, thanks. Done.

File: src/Function/Endomorphism/Propositional.agda
Comment (JacquesCarette) [reply src/Function/Endomorphism/Propositional.agda:101]: I've changed to the better name and deprecated the old one. And no to the instance: the types are quite different (you have to expanded them out to see it), so it's not immediate.

File: src/Function/Endomorphism/Setoid.agda
Comment (JacquesCarette) [reply src/Function/Endomorphism/Setoid.agda:102]: Done.

File: src/Algebra/Morphism/Structures.agda
Comment (jamesmckinna) [inline src/Algebra/Morphism/Structures.agda:71]: The `style-guide` indicates that there *should* be two blank lines after each internal `module`, although that rule is not observed rigidly throughout even this module, and in any case, arguably, the big section headings function more than adequately as separators... so not sure what to do here?

**Review by jamesmckinna — APPROVED** [review]
Otherwise all looking good, as per my earlier comments!

File: src/Algebra/Morphism/Structures.agda
Comment (JacquesCarette) [reply src/Algebra/Morphism/Structures.agda:71]: I think that big section headings are indeed more than adequate as separators - and we should update the style guide accordingly.

**Comment by JacquesCarette** [comment]
Could this get another review @MatthewDaggitt ?

File: src/Function/Endomorphism/Propositional.agda
Comment (MatthewDaggitt) [inline src/Function/Endomorphism/Propositional.agda:114]: So for this and the name below to be backwards compatible, we actually need the old names to be attached to the old definitions here, i.e. just copy the old definition wholesale.

File: src/Function/Endomorphism/Setoid.agda
Comment (MatthewDaggitt) [inline src/Function/Endomorphism/Setoid.agda:38]: Likewise here, we'd need to come up with an entirely new name for this (and indeed all the proofs below). That isn't really feasible, so I guess the best approach if you want to persue this is to simply leave the old module in place, and add an entirely new module under a new name with the new definitions....

**Review by MatthewDaggitt — COMMENTED** [review]
Sorry for abandoning this half way through the conversation!

**Comment by MatthewDaggitt** [comment]
Hmm so in summary, I think the only reasonable way of doing this in a backwards compatible manner would be to create entirely new modules (I would tentatively suggest `Function.Endomorphism.(Setoid/Propositional).New`?) with the new definitions. Then in `v3.0` we'll delete the current modules, and then switch the new ones in.

File: src/Function/Endomorphism/Setoid.agda
Comment (jamesmckinna) [reply src/Function/Endomorphism/Setoid.agda:38]: Hmmm... that seems... unfortunate.

**Comment by jamesmckinna** [comment]
Suggested alternative: Call the new module/modules `Endo.{Setoid|Propositional}`? (and avoid the awful `.New` subhierarchy?)

**Comment by JacquesCarette** [comment]
Ok, this now should be backwards compatible! If you could re-review @MatthewDaggitt and @jamesmckinna ?

File: src/Function/Endo/Propositional.agda
Comment (jamesmckinna) [inline src/Function/Endo/Propositional.agda:51]: As with the `TCO` versions in `Algebra.Definitions.RawMonoid`, should we consider such optimisations  here?

Indeed, should we in fact have `f ^ n` defined as in `Algebra.Properties.Monoid.Mult{.TCOptimised}`/`Algebra.Definitions.RawMonoid`, for the `RawMonoid` structure on `Endo`... and then the various homormor[hism properties etc. will follow from such definitions...?

File: src/Function/Endo/Setoid.agda
Comment (jamesmckinna) [inline src/Function/Endo/Setoid.agda:128]: Given that these (old) definitions now belong to entirely deprecated modules, and this module is *new*, is there any longer any need for these explicit deprecations?

File: src/Function/Endo/Propositional.agda
Comment (jamesmckinna) [inline src/Function/Endo/Propositional.agda:124]: Ditto. as with `Function.Endo.Setoid` below: now that we are in a *new* module, there's no need to explicitly export deprecated names which now exist only in deprecated modules, is there?

File: src/Function/Endo/Setoid.agda
Comment (jamesmckinna) [inline src/Function/Endo/Setoid.agda:34]: `style-guide`: Double blank line after initial `private` block, I believe...

**Review by jamesmckinna — CHANGES_REQUESTED** [review]
This all looks good, but have advised a number of refactorings, and removal of the (now-redundant!?) deprecations... and reinstatement of the #2277 `SuccessorSet` additions to `CHANGELOG`, which seem to have got deleted in the meantime...

**Comment by jamesmckinna** [comment]
Also @MatthewDaggitt : can we now remove `breaking` label?

File: CHANGELOG.md
Comment (JacquesCarette) [reply CHANGELOG.md:244]: I was afraid that would happen - I tried to do this via `git revert`, got into merge-hell over `CHANGELOG` and took in the simple-looking changes to this bit (because I couldn't believe that we wanted that complicated mess in the `CHANGELOG`...)  Will fix.

File: src/Function/Endo/Propositional.agda
Comment (JacquesCarette) [reply src/Function/Endo/Propositional.agda:51]: scope creep... but probably.

File: src/Function/Endo/Setoid.agda
Comment (JacquesCarette) [reply src/Function/Endo/Setoid.agda:128]: yes, definitely should have cleaned that up!

File: CHANGELOG.md
Comment (jamesmckinna) [reply CHANGELOG.md:244]: Well, sorry for it being a mess...;-) wasn't sure how much detail was required... but it did get merged the first time around, so I guess it had been approved...

File: src/Function/Endo/Propositional.agda
Comment (jamesmckinna) [reply src/Function/Endo/Propositional.agda:51]: Well... if you feel strongly, we can make that a downstream refactoring PR, but I think it's (eventually) what we should do?

Comment (JacquesCarette) [reply src/Function/Endo/Propositional.agda:51]: Yes, I think we should accept that this PR is enough of an improvement as is, and then make better still - later.

Comment (jamesmckinna) [reply src/Function/Endo/Propositional.agda:51]: Agreed. Happy to take that on downstream!

**Comment by JacquesCarette** [comment]
I think this now deals with all the comments from @jamesmckinna and @MatthewDaggitt and is no longer breaking either.

**Review by jamesmckinna — APPROVED** [review]

**Comment by jamesmckinna** [comment]
Looks good to me now! 
The HTML index run has failed... but hopefully that's a transient bug...

**Comment by JacquesCarette** [comment]
It's actually not - it is failing because of warnings. A deprecated module isn't allowed to use deprecated things lest warnings get issued. I had removed the warning suppression, sigh. Will have to restore it.

[This is probably something to file upstream - deprecated modules should be allowed to use other deprecated things, there is no need for a cascade of warnings!]

**Comment by jamesmckinna** [comment]
> It's actually not - it is failing because of warnings. A deprecated module isn't allowed to use deprecated things lest warnings get issued. I had removed the warning suppression, sigh. Will have to restore it.
> 
> [This is probably something to file upstream - deprecated modules should be allowed to use other deprecated things, there is no need for a cascade of warnings!]

Two or three reactions to this:
* suppressing warnings in a deprecated module seems harmless, if annoying that it should prove necessary
* that it now does so is/seems to be a knock-on consequence of #2224 although the check on deprecated modules claims to ignore warnings, so is the list broken in some way?
* maybe a clue to have (yet) another go at overhauling our CI ...?

File: src/Function/Endo/Propositional.agda
Comment (jamesmckinna) [inline src/Function/Endo/Propositional.agda:93]: Question in my mind as to whether these should be exported at top-level?
But certainly not a deal-breaker for this PR!

File: src/Function/Endo/Propositional.agda
Comment (jamesmckinna) [inline src/Function/Endo/Propositional.agda:102]: Ditto.: should `isRelHomomorphism` be lifted out here, too?

File: src/Function/Endo/Propositional.agda
Comment (jamesmckinna) [inline src/Function/Endo/Propositional.agda:61]: Do you *really* intend `_≡_` as the equality on `Endo`, rather than `_≗_`?

**Comment by jamesmckinna** [comment]
Re: #2394 See [this branch](https://github.com/jamesmckinna/agda-stdlib/tree/Endomorphism-bis/src/Function/Endo) for the details: easily incorporated on top of what you have here!

File: src/Function/Endo/Propositional.agda
Comment (jamesmckinna) [reply src/Function/Endo/Propositional.agda:61]: I realise that this is a legacy design decision, but given the green field reimplementation in this PR, a chance to reconsider?

Comment (JacquesCarette) [reply src/Function/Endo/Propositional.agda:61]: Agree, this is probably a mistake.

File: src/Function/Endo/Propositional.agda
Comment (JacquesCarette) [reply src/Function/Endo/Propositional.agda:93]: I guess since we're now moving to using `using` a lot, having extra names around isn't so costly, so I can make these visible beyond the bundle.

**Comment by JacquesCarette** [comment]
I think the names are fine as they are (for the purposes of porting this). Made minor tweaks, but I think once the build comes back green, it can be merged.

**Pull request merged**
